src/common/paddle/organization.ts

```diff
@@ -0,0 +1,120 @@
+import type {
+  SubscriptionCreatedEvent,
+  SubscriptionCreatedNotification,
+} from '@paddle/paddle-node-sdk';
+import { randomUUID } from 'crypto';
+import createOrGetConnection from '../../db';
+import { extractSubscriptionCycle, paddleInstance } from '.';
+import { Organization, User } from '../../entity';
+import { ContentPreferenceOrganization } from '../../entity/contentPreference/ContentPreferenceOrganization';
+import { ContentPreferenceStatus } from '../../entity/contentPreference/types';
+import { OrganizationMemberRoles } from '../../roles';
+import { SubscriptionProvider, SubscriptionStatus } from '../plus';
+import { logger } from '../../logger';
+import { isPlusMember } from '../../paddle';
+import { updateSubscriptionFlags } from '../utils';
+
+type PaddleCustomData = {
+  user_id?: string;
+};
+
+// If the user provides business information during the checkout, we will use it
+// otherwise we will use the default business name
+const getBusinessName = async (
+  user: User,
+  data: SubscriptionCreatedNotification,
+) => {
+  const defaultBusinessName = `${user.name}'s new Organization`;
+  if (!data.businessId) {
+    return defaultBusinessName;
+  }
+
+  try {
+    const business = await paddleInstance.businesses.get(
+      data.customerId,
+      data.businessId,
+    );
+    return business.name;
+  } catch (error) {
+    return defaultBusinessName;
+  }
+};
+
+export const createOrganizationSubscription = async ({
+  event,
+}: {
+  event: SubscriptionCreatedEvent;
+}) => {
+  const organizationId = randomUUID();
+  const con = await createOrGetConnection();
+
+  const { data } = event;
+  const { user_id: userId }: PaddleCustomData = data?.customData || {};
+
+  const subscriptionType = extractSubscriptionCycle(data.items);
+  if (!subscriptionType) {
+    logger.error(
+      {
+        provider: SubscriptionProvider.Paddle,
+        data: event,
+      },
+      'Subscription type missing in payload',
+    );
+    return false;
+  }
+
+  const user = await con.getRepository(User).findOneByOrFail({
+    id: userId,
+  });
+
+  const businessName = await getBusinessName(user, data);
+
+  const subscriptionFlags = {
+    cycle: subscriptionType,
+    createdAt: data.startedAt ?? undefined,
+    subscriptionId: data.id,
+    provider: SubscriptionProvider.Paddle,
+    status: SubscriptionStatus.Active,
+  };
+
+  const organization = await con.getRepository(Organization).save({
+    id: organizationId,
+    name: businessName,
+    seats: data.items[0].quantity,
+    subscriptionFlags: subscriptionFlags,
+  });
```

Comment:
Just checking if we want to run all logic in here or rather in a background script via topic (user-organisation-created) kind of thing?

Regardless maybe wrap everything here in a transaction? (or is con a passed transaction, can't remember)

Reply:
Probably should wrap it all in a transaction indeed. Will do that

Reply:
we could do it async as well, but transaction is a good start